File: src/lib/shoppingLocal.ts
import type { ShoppingItem } from "./types";

/** localStorage-backed shopping list used in Demo Mode. */

const KEY = "adaptable.shopping.v1";

function load(): ShoppingItem[] {
  try {
    const raw = localStorage.getItem(KEY);
    if (raw) return JSON.parse(raw) as ShoppingItem[];
  } catch {
    /* corrupted — start fresh */
  }
  return [];
}

let items = load();

function persist() {
  try {
    localStorage.setItem(KEY, JSON.stringify(items));
  } catch {
    /* storage unavailable — keep in memory */
  }
}

function newId(): string {
  return typeof crypto !== "undefined" && "randomUUID" in crypto
    ? crypto.randomUUID()
    : `it-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
}

export const shoppingLocal = {
  list(): ShoppingItem[] {
    return [...items];
  },
  add(
    rows: Array<Pick<ShoppingItem, "recipe_id" | "recipe_title" | "item" | "quantity">>,
  ): ShoppingItem[] {
    const created = rows.map((r) => ({
      ...r,
      id: newId(),
      checked: false,
      created_at: new Date().toISOString(),
    }));
    items = [...created, ...items];
    persist();
    return created;
  },
  setChecked(id: string, checked: boolean) {
    items = items.map((i) => (i.id === id ? { ...i, checked } : i));
    persist();
  },
  remove(id: string) {
    items = items.filter((i) => i.id !== id);
    persist();
  },
  clearChecked() {
    items = items.filter((i) => !i.checked);
    persist();
  },
};

File: ios/Adaptable/Adaptable/Services/ShoppingLocal.swift
import Foundation

/// UserDefaults-backed shopping list used in Demo Mode. Mirrors
/// `src/lib/shoppingLocal.ts`.
@MainActor
final class ShoppingLocal {
    static let shared = ShoppingLocal()

    private let key = "adaptable.shopping.v1"
    private var items: [ShoppingItem]

    private init() {
        if let raw = UserDefaults.standard.data(forKey: "adaptable.shopping.v1"),
           let decoded = try? JSONDecoder().decode([ShoppingItem].self, from: raw) {
            items = decoded
        } else {
            items = []
        }
    }

    private func persist() {
        if let data = try? JSONEncoder().encode(items) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }

    func list() -> [ShoppingItem] { items }

    func add(_ rows: [(recipeId: String?, recipeTitle: String, item: String, quantity: String)]) -> [ShoppingItem] {
        let now = ISO8601DateFormatter().string(from: Date())
        let created = rows.map { row in
            ShoppingItem(
                id: "it-\(Int(Date().timeIntervalSince1970 * 1000))-\(Int.random(in: 1000...9999))",
                recipe_id: row.recipeId,
                recipe_title: row.recipeTitle,
                item: row.item,
                quantity: row.quantity,
                checked: false,
                created_at: now
            )
        }
        items = created + items
        persist()
        return created
    }

    func setChecked(_ id: String, checked: Bool) {
        items = items.map {
            var i = $0
            if i.id == id { i.checked = checked }
            return i
        }
        persist()
    }

    func remove(_ id: String) {
        items.removeAll { $0.id == id }
        persist()
    }

    func clearChecked() {
        items.removeAll { $0.checked }
        persist()
    }
}
